types/src/transaction/authenticator.rs

```diff
@@ -300,6 +297,52 @@ impl TransactionAuthenticator {
             } => Some(fee_payer_signer.clone()),
         }
     }
+
+    pub fn all_signers(&self) -> Vec<AccountAuthenticator> {
+        match self {
+            // This is to ensure that any new TransactionAuthenticor variant must update this function.
+            Self::Ed25519 { .. }
+            | Self::MultiEd25519 { .. }
+            | Self::MultiAgent { .. }
+            | Self::FeePayer { .. }
+            | Self::SingleSender { .. }  => {
+                let mut account_authenticators: Vec<AccountAuthenticator> = vec![];
+                account_authenticators.push(self.sender());
+                account_authenticators.extend(self.secondary_signers());
+                if let Some(fee_payer) = self.fee_payer_signer() {
+                    account_authenticators.push(fee_payer);
+                }
+                account_authenticators
+            },
+        }
+    }
+
+    pub fn to_single_key_authenticators(&self) -> Result<Vec<SingleKeyAuthenticator>> {
+        let account_authenticators = self.all_signers();
+        let mut single_key_authenticators:Vec<SingleKeyAuthenticator> = Vec::with_capacity(MAX_NUM_OF_SIGS);
+        for account_authenticator in account_authenticators {
+            match account_authenticator {
+                AccountAuthenticator::Ed25519 { public_key, signature } => {
+                    let authenticator = SingleKeyAuthenticator{public_key: AnyPublicKey::ed25519(public_key.clone()), signature: AnySignature::ed25519(signature.clone())};
+                    single_key_authenticators.push(authenticator);
+                }
+                AccountAuthenticator::MultiEd25519 { public_key, signature } => {
+                    let public_keys = MultiKey::from(public_key);
+                    let signatures : Vec<AnySignature> = signature.signatures().iter().map(|sig| AnySignature::ed25519(sig.clone())).collect();
+                    let signatures_bitmap = aptos_bitvec::BitVec::from(signature.bitmap().to_vec());
+                    let authenticator = MultiKeyAuthenticator{public_keys, signatures, signatures_bitmap};
+                    single_key_authenticators.extend(authenticator.to_single_key_authenticators()?);
+                }
+                AccountAuthenticator::SingleKey { authenticator } => {
+                    single_key_authenticators.push(authenticator);
+                },
+                AccountAuthenticator::MultiKey { authenticator } => {
+                    single_key_authenticators.extend(authenticator.to_single_key_authenticators()?);
+                },
+            };
+        }
```

Comment:
nit:
```
single_key_authenticators: Vec<SingleKeyAuthenticator> = account_authenticators.iter().flat_map(AccountAuthenticator::as_single_key_authenticator).collect();
```
where `AccountAuthenticator::as_single_key_authenticator` is a new function.

Reply:
i tried but dealing with propagating the error from AccountAuthenticator::as_single_key_authenticator is annoying